File: backend/app/schemas/credit_notes.py
"""Tax credit note request/response schemas. Extra keys → 422."""

from datetime import date, datetime
from decimal import Decimal
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.credit_note import CreditNoteReason, CreditNoteStatus


class CreditNoteItemCreate(BaseModel):
    """Qty against a parent invoice line. Money fields optional; mismatch → 422."""

    model_config = ConfigDict(extra="forbid")

    invoice_item_id: UUID
    quantity: Decimal = Field(..., gt=0)
    unit_price: Optional[Decimal] = Field(None, ge=0)
    tax_rate: Optional[Decimal] = Field(None, ge=0, le=100)
    discount_percent: Optional[Decimal] = Field(None, ge=0)
    discount_amount: Optional[Decimal] = Field(None, ge=0)
    description: Optional[str] = Field(None, max_length=500)
    product_id: Optional[UUID] = None


class CreditNoteItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    credit_note_id: UUID
    invoice_item_id: UUID
    product_id: Optional[UUID] = None
    uom_id: Optional[UUID] = None
    sku_snapshot: Optional[str] = None
    description: str
    quantity: Decimal
    unit_price: Decimal
    tax_rate: Decimal
    discount_percent: Decimal
    discount_amount: Decimal
    line_net: Decimal
    tax_amount: Decimal
    total_price: Decimal
    created_at: datetime
    updated_at: datetime


class CreditNoteCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    invoice_id: UUID
    reason: CreditNoteReason
    reason_notes: Optional[str] = Field(None, max_length=5000)
    issue_date: Optional[date] = None
    items: List[CreditNoteItemCreate] = Field(..., min_length=1)


class CreditNoteUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    reason: Optional[CreditNoteReason] = None
    reason_notes: Optional[str] = Field(None, max_length=5000)
    issue_date: Optional[date] = None
    items: Optional[List[CreditNoteItemCreate]] = Field(None, min_length=1)


class CreditNoteResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    workspace_id: UUID
    client_id: UUID
    invoice_id: UUID
    credit_note_number: str
    status: CreditNoteStatus
    currency: str
    issue_date: date
    reason: CreditNoteReason
    reason_notes: Optional[str] = None
    subtotal: Decimal
    tax_amount: Decimal
    total_amount: Decimal
    original_invoice_number: Optional[str] = None
    original_issue_date: Optional[date] = None
    invoice_kind: Optional[str] = None
    seller_trn_snapshot: Optional[str] = None
    seller_name_snapshot: Optional[str] = None
    seller_address_snapshot: Optional[str] = None
    buyer_trn_snapshot: Optional[str] = None
    buyer_name_snapshot: Optional[str] = None
    buyer_address_snapshot: Optional[str] = None
    items: List[CreditNoteItemResponse] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime
    deleted_at: Optional[datetime] = None


class CreditNoteListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    client_id: UUID
    invoice_id: UUID
    credit_note_number: str
    status: CreditNoteStatus
    total_amount: Decimal
    issue_date: date
    created_at: datetime


class CreditNoteIssueRequest(BaseModel):
    """Empty issue body. Extra keys → 422."""

    model_config = ConfigDict(extra="forbid")

File: frontend/src/api/creditNotes.ts
import { apiClient } from './client';
import type { PaginatedResponse, PaginationMeta, SuccessResponse } from '../types/api';

export type CreditNoteStatus = 'DRAFT' | 'ISSUED';

export type CreditNoteReason =
  | 'SALES_RETURN'
  | 'INVOICE_ERROR'
  | 'DISCOUNT'
  | 'GOODWILL'
  | 'OTHER';

export interface CreditNoteItemWrite {
  invoice_item_id: string;
  quantity: number;
}

export interface CreditNoteItem {
  id: string;
  credit_note_id: string;
  invoice_item_id: string;
  product_id?: string | null;
  uom_id?: string | null;
  sku_snapshot?: string | null;
  description: string;
  quantity: number | string;
  unit_price: number | string;
  tax_rate: number | string;
  discount_percent: number | string;
  discount_amount: number | string;
  line_net: number | string;
  tax_amount: number | string;
  total_price: number | string;
  created_at?: string;
  updated_at?: string;
}

export interface CreditNoteListItem {
  id: string;
  client_id: string;
  invoice_id: string;
  credit_note_number: string;
  status: CreditNoteStatus;
  total_amount: number | string;
  issue_date: string;
  created_at?: string;
}

export interface CreditNote extends CreditNoteListItem {
  workspace_id?: string;
  currency?: string;
  reason: CreditNoteReason;
  reason_notes?: string | null;
  subtotal: number | string;
  tax_amount: number | string;
  original_invoice_number?: string | null;
  original_issue_date?: string | null;
  invoice_kind?: string | null;
  seller_trn_snapshot?: string | null;
  seller_name_snapshot?: string | null;
  seller_address_snapshot?: string | null;
  buyer_trn_snapshot?: string | null;
  buyer_name_snapshot?: string | null;
  buyer_address_snapshot?: string | null;
  items: CreditNoteItem[];
  updated_at?: string;
  deleted_at?: string | null;
}

export interface CreditNoteCreatePayload {
  invoice_id: string;
  reason: CreditNoteReason;
  reason_notes?: string;
  issue_date?: string;
  items: CreditNoteItemWrite[];
}

export interface CreditNoteUpdatePayload {
  reason?: CreditNoteReason;
  reason_notes?: string | null;
  issue_date?: string;
  items?: CreditNoteItemWrite[];
}

export interface CreditNoteListQuery {
  page?: number;
  per_page?: number;
  status?: CreditNoteStatus;
  client_id?: string;
  invoice_id?: string;
  search?: string;
}

export interface CreditNoteListResult {
  items: CreditNoteListItem[];
  pagination: PaginationMeta;
}

const BASE = '/api/v1/credit-notes';

function listParams(query: CreditNoteListQuery): Record<string, string | number> {
  const params: Record<string, string | number> = {
    page: query.page ?? 1,
    per_page: query.per_page ?? 20,
  };
  if (query.status) params.status = query.status;
  if (query.client_id) params.client_id = query.client_id;
  if (query.invoice_id) params.invoice_id = query.invoice_id;
  if (query.search?.trim()) params.search = query.search.trim();
  return params;
}

function unwrapList(payload: PaginatedResponse<CreditNoteListItem>): CreditNoteListResult {
  const items = Array.isArray(payload.data) ? payload.data : [];
  return { items, pagination: payload.pagination };
}

function optionalText(value?: string | null): string | undefined {
  const trimmed = value?.trim();
  return trimmed || undefined;
}

function itemBody(item: CreditNoteItemWrite): Record<string, unknown> {
  return { invoice_item_id: item.invoice_item_id, quantity: item.quantity };
}

export function createCreditNoteBody(
  data: CreditNoteCreatePayload,
): Record<string, unknown> {
  const body: Record<string, unknown> = {
    invoice_id: data.invoice_id,
    reason: data.reason,
    items: data.items.map(itemBody),
  };
  const notes = optionalText(data.reason_notes);
  if (notes) body.reason_notes = notes;
  if (data.issue_date) body.issue_date = data.issue_date;
  return body;
}

export function updateCreditNoteBody(
  data: CreditNoteUpdatePayload,
): Record<string, unknown> {
  const body: Record<string, unknown> = {};
  if (data.reason) body.reason = data.reason;
  if (data.reason_notes !== undefined) {
    body.reason_notes = optionalText(data.reason_notes) ?? null;
  }
  if (data.issue_date) body.issue_date = data.issue_date;
  if (data.items && data.items.length > 0) {
    body.items = data.items.map(itemBody);
  }
  return body;
}

export const getCreditNotes = async (
  query: CreditNoteListQuery = {},
): Promise<CreditNoteListResult> => {
  const response = await apiClient.get<PaginatedResponse<CreditNoteListItem>>(BASE, {
    params: listParams(query),
  });
  return unwrapList(response.data);
};

export const getCreditNote = async (id: string): Promise<CreditNote> => {
  const response = await apiClient.get<SuccessResponse<CreditNote>>(`${BASE}/${id}`);
  return response.data.data;
};

export const createCreditNote = async (
  data: CreditNoteCreatePayload,
): Promise<CreditNote> => {
  const response = await apiClient.post<SuccessResponse<CreditNote>>(
    BASE,
    createCreditNoteBody(data),
  );
  return response.data.data;
};

export const updateCreditNote = async (
  id: string,
  data: CreditNoteUpdatePayload,
): Promise<CreditNote> => {
  const response = await apiClient.put<SuccessResponse<CreditNote>>(
    `${BASE}/${id}`,
    updateCreditNoteBody(data),
  );
  return response.data.data;
};

export const deleteCreditNote = async (id: string): Promise<void> => {
  await apiClient.delete(`${BASE}/${id}`);
};

export const issueCreditNote = async (id: string): Promise<CreditNote> => {
  const response = await apiClient.post<SuccessResponse<CreditNote>>(
    `${BASE}/${id}/issue`,
    {},
  );
  return response.data.data;
};
